### PR TITLE
Topic/authorization

### PR DESCRIPTION
**Pull request by frehn**
This is a continuation of PR #264, adding HTTP Basic and Digest Authorization in the form of middleware. There is one open problem left concerning the composition of services using `orElse`. The solution of this problem was deferred to later; for details see the discussion in PR #264.

### CODE REVIEW

File: server/src/main/scala/org/http4s/server/middleware/authentication/DigestAuthentication.scala [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+  private case object BadParameters extends AuthReply
+
+  private def checkAuth(req: Request) = req.headers.get(Authorization) match {
+      case Some(Authorization(GenericCredentials(AuthScheme.Digest, params))) => checkAuthParams(req, params)
+      case Some(Authorization(_)) => Task.now(NoCredentials)
+      case None => Task.now(NoAuthorizationHeader)
+  }
+
+  private def getChallengeParams(staleNonce: Boolean) = Task {
```
Comment (julien-truffaut) [inline server/src/main/scala/org/http4s/server/middleware/authentication/DigestAuthentication.scala]: would you mind to specify the return type explicitly, it gives a better doc

File: server/src/main/scala/org/http4s/server/middleware/authentication/Authentication.scala [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  def apply(service: HttpService): HttpService = Service.lift {
+    case req: Request => getChallenge(req).flatMap(_ match {
+      case \/-(req) => service(req)
+      case -\/(challenge) =>
+        Task.now(Some(Response(Status.Unauthorized).putHeaders(`WWW-Authenticate`(challenge))))
+    })
+  }
+}
+
+abstract class AuthReply
```
Comment (julien-truffaut) [inline server/src/main/scala/org/http4s/server/middleware/authentication/Authentication.scala]: can we make `AuthReply` sealed?

**Comment by bryce-anderson** [comment]
@frehn, I'm going to merge this and quickly take care of @julien-truffaut's suggestions in a very quick follow up commit. Thanks a lot for this work, and sorry the merge process has been so long: typically it would be faster where it not for the problem we still face with our monad stack ordering.

Thanks again.

**Comment by frehn** [comment]
@bryce-anderson, thanks for merging. I very much enjoyed the discussion and feedback on the PR and am looking forward to contributing to http4s again in the future.

**Pull request merged**